src/NHibernate/Criterion/EntityProjection.cs

```diff
@@ -0,0 +1,175 @@
+
+		#region IProjection implementation
+
+		/// <summary>
+		/// Entity alias
+		/// </summary>
+		public string[] Aliases
+		{
+			get { return new[] {_entityAlias}; }
+		}
+
+		bool IProjection.IsAggregate
+		{
+			get { return false; }
+		}
+
+		bool IProjection.IsGrouped
+		{
+			get { return false; }
+		}
+
+		IType[] IProjection.GetTypes(string alias, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			return new[] {criteriaQuery.GetType(criteria, alias)};
+		}
+
+		IType[] IProjection.GetTypes(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _types;
+		}
+
+		public string[] GetColumnAliases(int position, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _columnAliases;
+		}
+
+		public string[] GetColumnAliases(string alias, int position, ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			return _columnAliases;
+		}
+
+		SqlString IProjection.ToSqlString(ICriteria criteria, int position, ICriteriaQuery criteriaQuery)
+		{
+			SetFields(criteria, criteriaQuery);
+
+			string identifierSelectFragment = Persister.IdentifierSelectFragment(_tableAlias, _columnAliasSuffix);
+			return new SqlString(
+				Lazy
+					? identifierSelectFragment
+					: string.Concat(
+						identifierSelectFragment,
+						Persister.PropertySelectFragment(_tableAlias, _columnAliasSuffix, FetchLazyProperties)));
+		}
+
+		SqlString IProjection.ToGroupSqlString(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			throw new NotImplementedException();
+		}
+
+		TypedValue[] IProjection.GetTypedValues(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			throw new NotImplementedException();
+		}
+
+		#endregion IProjection implementation
+
+		private void SetFields(ICriteria criteria, ICriteriaQuery criteriaQuery)
+		{
+			//Persister is required, so let's use it as "initialized marker"
+			if (Persister != null)
+				return;
+
+			if (RootEntity == null)
+			{
+				RootEntity = criteria.GetRootEntityTypeIfAvailable();
+			}
+
+			if (_entityAlias == null)
+			{
+				_entityAlias = criteria.Alias;
+			}
+
+			Persister = (IQueryable) criteriaQuery.Factory.GetEntityPersister(RootEntity.FullName);
+
+			ICriteria subcriteria = criteria.GetCriteriaByAlias(_entityAlias);
+			if (subcriteria == null)
+				throw new HibernateException($"Criteria\\QueryOver alias '{_entityAlias}' for entity projection is not found.");
+
+			_tableAlias = criteriaQuery.GetSQLAlias(
+				subcriteria,
+				Persister.IdentifierPropertyName ?? string.Empty);
+
+			_columnAliasSuffix = criteriaQuery.GetIndexForAlias().ToString();
+
+			IdentifierColumnAliases = Persister.GetIdentifierAliases(_columnAliasSuffix);
+
+			PropertyColumnAliases = Lazy
+				? new string[][] { }
+				: Enumerable.Range(0, Persister.PropertyNames.Length).Select(i => Persister.GetPropertyAliases(_columnAliasSuffix, i)).ToArray();
```

Comment:
Indexes for PropertyColumnAliases must be the same as in Persister.PropertyNames (it's used somewhere deep inside for property mapping - I forgot where)

So I guess I could only add some micro-optimization to avoid alias calculation for lazy properties and return empty string collection for it. If you think it's worth it.

Reply:
> So I guess I could only add some micro-optimization to avoid alias calculation for lazy properties and return empty string collection for it. If you think it's worth it.

No that is ok for me as this part of the PR is currently if it is already a constraint of current code base to always have lazy properties listed here.